aerospike-core/src/batch/batch_executor.rs

```diff
@@ -40,73 +40,39 @@ impl BatchExecutor {
-        }
-        Ok(res)
+        let mut all_results: Vec<_> = reads.into_iter().flat_map(|cmd|cmd.batch_reads).collect();
+        all_results.sort_by_key(|(_, i)|*i);
+        Ok(all_results.into_iter().map(|(b, _)|b).collect())
     }
 
     async fn execute_batch_jobs(
         &self,
         jobs: Vec<BatchReadCommand>,
         concurrency: &Concurrency,
     ) -> Result<Vec<BatchReadCommand>> {
-        let threads = match *concurrency {
-            Concurrency::Sequential => 1,
-            Concurrency::Parallel => jobs.len(),
-            Concurrency::MaxThreads(max) => cmp::min(max, jobs.len()),
-        };
-        let size = jobs.len() / threads;
-        let mut overhead = jobs.len() % threads;
-        let last_err: Arc<Mutex<Option<Error>>> = Arc::default();
-        let mut slice_index = 0;
-        let mut handles = vec![];
-        let res = Arc::new(Mutex::new(vec![]));
-        for _ in 0..threads {
-            let mut thread_size = size;
-            if overhead >= 1 {
-                thread_size += 1;
-                overhead -= 1;
-            }
-            let slice = Vec::from(&jobs[slice_index..slice_index + thread_size]);
-            slice_index = thread_size + 1;
-            let last_err = last_err.clone();
-            let res = res.clone();
-            let handle = aerospike_rt::spawn(async move {
-                //let next_job = async { jobs.lock().await.next().await};
-                for mut cmd in slice {
-                    if let Err(err) = cmd.execute().await {
-                        *last_err.lock().await = Some(err);
-                    };
-                    res.lock().await.push(cmd);
-                }
-            });
-            handles.push(handle);
-        }
-        futures::future::join_all(handles).await;
-        match Arc::try_unwrap(last_err).unwrap().into_inner() {
-            None => Ok(res.lock().await.to_vec()),
-            Some(err) => Err(err),
+        let handles = jobs.into_iter().map(BatchReadCommand::execute);
+        match concurrency {
+            Concurrency::Sequential => futures::future::join_all(handles).await.into_iter().collect(),
+            Concurrency::Parallel => futures::future::join_all(handles.map(aerospike_rt::spawn)).await.into_iter().map(|value|value.map_err(|e|e.to_string())?).collect(),
         }
     }
 
     async fn get_batch_nodes(
         &self,
         batch_reads: &[BatchRead],
-    ) -> Result<HashMap<Arc<Node>, Vec<BatchRead>>> {
+    ) -> Result<HashMap<Arc<Node>, Vec<(BatchRead, usize)>>> {
         let mut map = HashMap::new();
-        for (_, batch_read) in batch_reads.iter().enumerate() {
+        for (index, batch_read) in batch_reads.iter().enumerate() {
             let node = self.node_for_key(&batch_read.key).await?;
             map.entry(node)
                 .or_insert_with(Vec::new)
-                .push(batch_read.clone());
+                .push((batch_read.clone(), index));
```

Comment:
Previously batch reads would take indexes... changing it in the async patch introduced a bug where batches might return out of order with multiple nodes.

This retains the index so it can be resorted when written.